enhancements/machine-api/control-plane-machineset-vertical-scaling.md

```diff
@@ -0,0 +1,467 @@
+  machinesetup with a machinetype and concrete values that must be adjusted. The
+  implementation should use multiple [discriminated
+  unions](https://github.com/openshift/enhancements/blob/master/dev-guide/api-conventions.md#discriminated-unions).
+  One will specify the `machineType` used:
+  - `machineType`: currently always `machines_v1beta1_machine_openshift_io`, but
+  should be able to support [cluster API](https://cluster-api.sigs.k8s.io/) in
+  the future. 
+    - `machines_v1beta1_machine_openshift_io`: specifies options for the
+    machines API based provisioner.
+      - `machineConfigurations`: specifies the concrete specifications for the
+        machines.
+  
+  The `machineConfiguration` will use a discriminated union again to dispatch
+  for the cloud provider. The dispatch key will be `platform`. To start the
+  supported cloud providers must include:
+  - `Azure` and configuration for `vmSize`
+  - `GCP` and configuration for `machineType`
+  - `AWS` and configuration for `instanceType`
+  - `vSphere` and configuration for `memoryMiB`, `numCPUs` and `diskGiB`.
+  
+  Each element that denotes an instance configuration also has to include a
+  `weight` property to specify the order in which scaling should occur.
+
+The **optional** configurations will include the following properties:
+
+- `syncPeriod`: how often will the operator check resource usage to see if scaling
+  should be performed.
+- `scaleUp`: defines configuration for scaling up. This includes the following
+  sub elements:
+  - `stabilizationWindow`: duration to wait after a scale up before another
+    scaling operation can occur.
+  - `selectPolicy`: `next` to select the next bigger instance size.
+  - `triggerPolicy`: `all` or `any`. Determines if scaling occurs if one or
+    all triggers have to be true.
+  - `triggers`: sets up the list of triggers, which will trigger a scale up. The
+    configuration is based on [custom metrics autoscaling
+    configuration](https://docs.openshift.com/container-platform/4.13/nodes/cma/nodes-cma-autoscaling-custom-trigger.html)
+    and should be kept compatbile to make usage easier for users accustomed to
+    CMA. Specifying multiple triggers, the `triggerPolicy` will determine when a
+    scaleup is performed.
+    An empty list of triggers disables scale up.
+    - `type`: `prometheus`, `cpu` or `memory`.
+    - `Prometheus`:
+      - `serverAddress`: https://thanos-querier.openshift-monitoring.svc.cluster.local:9092 
+      - `metricName`: Specifies the name to identify the metric in the
+        `external.metrics.k8s.io` API.
+      - `threshold`: Specifies the value that triggers scaling. 
+      - `query`: Specifies the Prometheus query to use.
+      - `authModes`: Specifies the authentication method to use. Should support
+        at least *basic*, *bearer* and *tls* authentication.
+      - `ignoreNullValues`: false 
+      - `unsafeSsl`: Specifies whether the certificate check should be skipped.
```

Comment:
Is this definitely something we want to support? Typically the field name for this would be something like `insecureSkipTLSVerify`, might be nice to be consistent.

Shouldn't be a bool btw. Would prefer to see, if at all possible, the ability to specific a CA certificate to verify against rather than allowing to skip verification completely

Reply:
It's coped from the CMA operator (https://keda.sh/docs/2.12/scalers/prometheus/#authentication-parameters) but I can also adjust it, if we are fine not having a 1:1 match to the fields in some cases.

Reply:
Yeah I don't think we need to match their API, looks like it hasn't been through API review so I would prefer we stick to OpenShift conventions rather than some third party project